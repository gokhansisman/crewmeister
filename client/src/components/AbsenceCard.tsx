import { Absence, Member } from "../utils/types";
import Moment from "react-moment";
import {  MdPending, MdCancel } from "react-icons/md";
import { FcCalendar } from "react-icons/fc";
import { AiFillCheckCircle } from "react-icons/ai";
import {
  Container,
  Row,
  CardHeader,
  CardHeaderRow,
  AddToCalendar,
  Profile,
  Columns,
  Note,
  NoteContainer,
  NoteOwner,
  Period,
  DateRow,
  VacationBadge,
  DateContainer,
  ProfileContainer,
  ProfileCard,
} from "./CardStyles";

export type AbsenceCardProps = {
  absence: Absence;
  member: Member;
};

const addToCalendar = async (absence: Absence, member: Member) => {
  const settings: RequestInit = {
    method: "POST",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      startTime: absence.startDate,
      endTime: absence.endDate,
      name: member.name,
      summary: absence.type,
    }),
  };
  try {
    const fetchResponse = await fetch(
      `/addToCalendar`,
      settings
    );
    if (await fetchResponse.ok) {
      const downloadICS = document.createElement("a");
      downloadICS.href = "/calendar";
      downloadICS.click();
    }
    return fetchResponse;
  } catch (e) {
    console.log(e);
  }
};
/**
 * AbsenceCard component renders the absence information of a member
 * @param {AbsenceCardProps} props
 * @returns {JSX.Element}
 */
function AbsenceCard({ absence, member }: AbsenceCardProps) {
  return (
    <Container>
      {absence.rejectedAt && (
        <Row>
          <CardHeader backgroundColor="#eb070766">
            <CardHeaderRow>
              <MdCancel size={40} color="#f44336" />
            </CardHeaderRow>
          </CardHeader>
          <VacationBadge shadowColor="#eb070766">
          {absence.type === "vacation" ? <>&#x2708;&#xFE0F;</> : <>&#x1F927;</>}
        </VacationBadge>
        </Row>
      )}
      {absence.confirmedAt && (
        <Row>
          <CardHeader backgroundColor="#00b57a9e">
            <CardHeaderRow>
              <AiFillCheckCircle size={40} color="#4caf50" />
            </CardHeaderRow>
          </CardHeader>
          <VacationBadge shadowColor="#00b57a9e">
          {absence.type === "vacation" ? <>&#x2708;&#xFE0F;</> : <>&#x1F927;</>}
        </VacationBadge>
        </Row>
        
      )}
      {!absence.confirmedAt && !absence.rejectedAt && absence.createdAt && (
        <Row>
          <CardHeader backgroundColor="#BCD1FF">
            <CardHeaderRow>
              <MdPending size={40} color="blue" />
            </CardHeaderRow>
          </CardHeader>
          <VacationBadge shadowColor="#BCD1FF">
          {absence.type === "vacation" ? <>&#x2708;&#xFE0F;</> : <>&#x1F927;</>}
        </VacationBadge>
        </Row>
      )}
      <DateContainer>
        <DateRow color="red">
          <Note>{absence.startDate + " / " + absence.endDate}</Note>
        </DateRow>
        <Period>
          <Moment diff={absence.startDate} unit="days">
            {absence.endDate}
          </Moment>
          days
        </Period>
      </DateContainer>
      {absence.admitterNote && (
        <NoteContainer>
          <NoteOwner>Admitter note</NoteOwner>
          {absence.admitterNote}
        </NoteContainer>
      )}
      {absence.memberNote && (
        <NoteContainer>
          <NoteOwner>Member note</NoteOwner>
          {absence.memberNote}
        </NoteContainer>
      )}
      <ProfileContainer>
        <ProfileCard>
          <Profile src={member && member.image} alt={member && member.name} />
          {member && member.name}
        </ProfileCard>
        <AddToCalendar onClick={() => addToCalendar(absence, member)}>
          <FcCalendar size={24} />
        </AddToCalendar>
      </ProfileContainer>
    </Container>
  );
}

export default AbsenceCard;
