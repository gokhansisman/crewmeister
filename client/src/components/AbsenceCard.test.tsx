import { shallow } from "enzyme";
import AbsenceCard from "./AbsenceCard";

const mockAbsence = {
  admitterId: 1,
  admitterNote: "",
  confirmedAt: new Date(2020 - 12 - 12),
  createdAt: new Date(2020 - 12 - 15),
  crewId: 352,
  endDate: "2021-01-13",
  id: 2351,
  memberNote: "",
  rejectedAt: new Date(),
  startDate: "2021-01-13",
  type: "sickness",
  userId: 2664,
};

const mockMember = {
  crewId: 352,
  id: 709,
  image: "https://loremflickr.com/300/400",
  name: "Max",
  userId: 644,
};

it("expect to render AbsenceCard component", () => {
  expect(
    shallow(<AbsenceCard absence={mockAbsence} member={mockMember} />).length
  ).toEqual(1); 
});
