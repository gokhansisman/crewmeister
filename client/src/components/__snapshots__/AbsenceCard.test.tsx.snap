// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect to take snapshot of AbsenceCard component 1`] = `ShallowWrapper {}`;
